helix-core/src/syntax.rs

```diff
@@ -1809,24 +1824,65 @@ impl<'a> Iterator for HighlightIter<'a> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum InjectionCapture<'a> {
+    Name(Cow<'a, str>),
+    File(Cow<'a, Path>),
+    Shebang(String),
+}
+
+static SHEBANG: &str = r"#!\s*(?:\S*[/\\](?:env\s+(?:\-\S+\s+)*)?)?([^\s\.\d]+)";
+
 fn injection_for_match<'a>(
     config: &HighlightConfiguration,
     query: &'a Query,
     query_match: &QueryMatch<'a, 'a>,
     source: RopeSlice<'a>,
-) -> (Option<Cow<'a, str>>, Option<Node<'a>>, IncludedChildren) {
+) -> (
+    Option<InjectionCapture<'a>>,
+    Option<Node<'a>>,
+    IncludedChildren,
+) {
     let content_capture_index = config.injection_content_capture_index;
     let language_capture_index = config.injection_language_capture_index;
+    let language_filename_index = config.injection_filename_capture_index;
+    let language_shebang_index = config.injection_shebang_capture_index;
 
-    let mut language_name = None;
+    let mut injection_capture = None;
     let mut content_node = None;
     for capture in query_match.captures {
         let index = Some(capture.index);
         if index == language_capture_index {
             let name = byte_range_to_str(capture.node.byte_range(), source);
-            language_name = Some(name);
+            injection_capture = Some(InjectionCapture::Name(name));
         } else if index == content_capture_index {
             content_node = Some(capture.node);
+        } else if index == language_filename_index {
+            let name = byte_range_to_str(capture.node.byte_range(), source);
+            let path = Path::new(name.as_ref()).to_path_buf();
+
+            injection_capture = Some(InjectionCapture::File(path.into()));
+        } else if index == language_shebang_index {
+            let line_idx = source.byte_to_line(capture.node.start_byte());
+
+            // some languages allow space and newlines before the actual string content
+            // so a shebang could be on either the first or second line
+            let lines = if source.len_lines() > line_idx + 2 {
+                source.slice(
+                    // start of first line
+                    source.line_to_char(line_idx)
+                    // end of second line
+                    ..source.line_to_char(line_idx + 2) - 1,
+                )
+            } else {
+                source.line(line_idx)
+            };
```

Comment:
Is it possible for this to look outside of the text of captured node? We might want to slice `source` to the byte range of `capture.node` and work from there

Reply:
I fixed this in another [commit](https://github.com/helix-editor/helix/pull/3970/commits/1541bd683dce26fef0d4b5dc66e160a3330181cf) that refactors the shebang function into a trait method to make the flow easier to follow.